[theman-oj10] iP

theman-oj10 opened the pull request:
# Alfred
> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))

Alfred frees your mind of having to remember things you need to do. It's,

- text-based
- easy to learn
- ~FAST~ SUPER FAST to use

All you need to do is,

1. download it from [here](https://nus-cs2103-ay2324s2.github.io/website/schedule/week4/project.html).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉

And it is **FREE**!

Features:

- [ ]  Managing tasks
- [x]  Managing deadlines (coming soon)
- [ ]  Reminders (coming soon)

If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:

```
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

theman-oj10 committed:
Duke Class: change name and initial greet of Duke

Duke class has some default greet behaviour and a default name.

The default behaviour and name do not fit the project requirements.

Changing the default behaviour and name will make the class adhere to project requirements.

Let's
* Change the name to Alfred
* Add the Initial greet as per project requirements

theman-oj10 committed:
Duke Class: implement Echo Functionality

Duke class only had a Greet and Exit functionality.

Added Echo functionality to fit the project requirements.

Adding Echo Functionality will make the class adhere to project requirements.

Let's
* Add Echo method
* Abstract the existing greet and bye functionalities into methods for better code quality

theman-oj10 committed:
Duke Class: implement list functionality

Duke class only had a Greet, Exit and Echo functionality.

Added List functionality to fit the project requirements.

Adding List Functionality will make the class adhere to project requirements.

Let's
* Implement a list to keep track of user inputs
* Modify Echo method to echo out user input that was added to the list
* Create printList method to print the list upon user's request

theman-oj10 committed:
Duke Class: implement Mark as Done, unmark functionality

Duke class only had a basic to do list functionality.

Improved to do List by creating a Task Class to fit the project requirements.

Adding Task Class will improve code quality and readability.

Let's
* Implement a Task Class
* Update existing Duke class to incorporate Task Class

theman-oj10 committed:
Alfred Class: Create new subclasses for Tasks and their corresponding functionalities

There are different kinds of tasks each with their unique kind of functionalities.

Creating these subclasses allows Alfred to be more robust and handle the different tasks as per the project requirements. Renaming Classes and Refactoring classes also improves code quality and readability.

Let's
* Implement a Todo, Event, Deadline subclasses
* Rename Duke class to Alfred
* Create a new package for seamless importing of classes

theman-oj10 committed:
Create semi-automatic testing for core functionality

All testing is manual so far.

Creating semi-automatic testing scripts using input/output redirection improves code quality and reduces bugs.

Let's
* Update the given bash script
* Create a custom test case

theman-oj10 committed:
Implement error handing for most user input cases

There are many invalid user inputs that the Alfred class cannot deal with.

Implementing Error handling improves code quality and functionality. Suitable error messages will instruct the user on how to correctly use the chatbot.

Let's
* Implement Alfred Exception Class
* Identify all possible invalid inputs and implement suitable error messages

theman-oj10 committed:
Alfred Class: Implement Delete functionality for the to-do list

Alfred class is able to add, mark, unmark, view the to-do list but unable to remove entries from the list.

Implementing this feature will improve usability of the chatbot and will make it adhere to project requirements.

Let's
* Implement deleteList method and appropriate error handling
* Implement error handling for markList and unmarkList which was missed out in the previous commit

theman-oj10 committed:
Alfred Class: Implement TaskTypes using Enums

Alfred class uses three types of Tasks: Todos, Deadlines and Events.

Using Enums will prevent accidental string comparisons that might lead to unpredictable behaviour.

Let's
* Implement TaskType Enum
* Update addList method to use TaskType
* Abstract out repeated functionality to reduce code duplication

wallacexuhanxiao commented on src/main/java/chatbot/Alfred.java:
```diff
@@ -6,18 +6,15 @@
 public class Alfred {
     protected static final String line = "________________________________________________________________________________________";
```
Is it better if we use ```LINE``` rather than ```line``` for **CONSTANT**s?

wallacexuhanxiao commented on src/main/java/chatbot/Alfred.java:
```diff
@@ -6,18 +6,15 @@
 public class Alfred {
     protected static final String line = "________________________________________________________________________________________";
     protected static ArrayList<Task> list = new ArrayList<Task>();
```
Maybe a more intuitive name like ```tasks``` should be better?

wallacexuhanxiao commented on src/main/java/chatbot/Alfred.java:
```diff
@@ -26,50 +23,91 @@ public void echo(String input) {
-                list.add(new Event(description, startTime, endTime));
-            } else {
+                input = input.substring(6).trim();
+            }
```
Too much ```if``` ```else```, maybe abstact it into a function?

wallacexuhanxiao commented on src/main/java/chatbot/Alfred.java:
```diff
@@ -78,11 +116,10 @@ public void addList(String input) {
         this.echo(String.format("Got it. I've added this task:\n  %s\nNow you have %d %s in the list.", list.get(list.size()-1).toString(), list.size(), singularTask));
```
Should there be a space before and after ```-``` ?

wallacexuhanxiao commented on src/main/java/chatbot/Alfred.java:
```diff
@@ -93,120 +130,62 @@ public void printList() {
     public void unmarkList(String index) {
-        int extractedIdx = 0;
-        try {
-        if (index.isEmpty()) {
```
Maybe the indentation here is wrong? 🤔

wallacexuhanxiao commented on src/main/java/chatbot/Alfred.java:
```diff
@@ -93,120 +130,62 @@ public void printList() {
+        list.remove(extractedIdx);
+        this.echo("Noted. I've removed this task:\n  " + removedTask + "\nNow you have " + list.size() + " tasks in the list.");
+    }
+    private int validateAndExtractIndex(String index) {
```
Maybe it is a good idea to put some of the codes in this function into another helper function? 🤔 too much ```try``` ```if``` stick together.

theman-oj10 replied:
validateAndExtractIndex is already a helper function

wallacexuhanxiao commented on src/main/java/chatbot/Alfred.java:
```diff
@@ -93,120 +130,62 @@ public void printList() {
+        return -1;
     }
 
     public static void main(String[] args) {
```
I like this ```main``` function. 👍

wallacexuhanxiao commented on src/main/java/chatbot/Alfred.java:
```diff
@@ -26,50 +23,91 @@ public void echo(String input) {
+                    }
+                    list.add(new Event(descriptionEvent, startTime, endTime));
+                    break;
+            }
```
Should we add a ```default``` for the ```switch``` sentences?

wallacexuhanxiao reviewed (commented):
LGTM. Just some nits to fix. 👍

theman-oj10 committed:
Alfred Class: Implement Save to Hard Disk Feature

Alfred class creates a new todo list every time it is run and data is lost when it is exited.

Using local storage to store the data will allow an user to continue from where he or she left of improving the utility of the application.

Let's
* Implement addData method
* Implement processData method
* Implement error handling to create a new data file if data file is lost

theman-oj10 committed:
Merge branch 'branch-Level-7'
Merge Commit: Merging Save Feature from level-7 into the master branch

Tanzhiheng26 commented on src/main/java/chatbot/Alfred.java:
```diff
@@ -0,0 +1,315 @@
+                        throw new AlfredException("Sorry Master Bruce. Please specify the due date or time by including /by due-date.");
+                    }
+                    Deadline deadline = new Deadline(description, by);
+                    //addData(deadline);
```
Should this be removed?

Tanzhiheng26 commented on src/main/java/chatbot/Alfred.java:
```diff
@@ -0,0 +1,315 @@
+        try {
+            java.io.FileWriter fw = new java.io.FileWriter("data/alfred.txt", true);
+            String taskType = task.getClass().getName();
+            switch (taskType) {
+                case "chatbot.ToDo":
+                    fw.write("T | " + (task.isDone() ? 1 : 0) + " | " + task.getDescription() + "\n");
+                    break;
+                case "chatbot.Deadline":
+                    Deadline deadline = (Deadline) task;
+                    String by = deadline.getBy();
+                    String[] splitResult = by.split(" ");
+                    // Join the parts with |
+                    String formattedBy = String.join(" | ", splitResult);
+                    fw.write("D | " + (task.isDone() ? 1 : 0) + " | " + task.getDescription() + " | " + formattedBy + "\n");
+                    break;
+
+                case "chatbot.Event":
+                    Event event = (Event) task;
+                    String[] splitResults = event.getStartTime().split(" ", 3);
+                    String[] splitResults2 = event.getEndTime().split(" ", 3);
+                    String startFormattedBy = String.join(" | ", splitResults);
+                    String endFormattedBy = String.join(" | ", splitResults2);
+                    fw.write("E | " + (task.isDone() ? 1 : 0) + " | " + task.getDescription() + " | " + startFormattedBy + " | "+ endFormattedBy + "\n");
+                    break;
```
Case clauses should not be indented?

Tanzhiheng26 commented on src/main/java/chatbot/Alfred.java:
```diff
@@ -0,0 +1,315 @@
+                        throw new AlfredException("Sorry Master Bruce. Please specify the start time and end time by including /from start-time /to end-time.");
+                    }
+                    Event event = new Event(descriptionEvent, startTime, endTime);
+                    //addData(event);
```
Should this be removed?

Tanzhiheng26 commented on src/main/java/chatbot/Alfred.java:
```diff
@@ -0,0 +1,315 @@
+            e.printError();
+        }
+        String singularTask = list.size() == 1 ? "task" : "tasks";
+        this.echo(String.format("Got it. I've added this task:\n  %s\nNow you have %d %s in the list.", list.get(list.size()-1).toString(), list.size(), singularTask));
```
Line is longer than 120 characters

Tanzhiheng26 commented on src/main/java/chatbot/Alfred.java:
```diff
@@ -0,0 +1,315 @@
+                if (list.size() == 1) {
+                    throw new AlfredException("Sorry Master Bruce. The task number you have entered is not in the list. There is only one item in the list.");
+                } else {
+                    throw new AlfredException("Sorry Master Bruce. The task number you have entered is not in the list. Please enter a number in the range of 1 to " + list.size() + ".");
```
Line is longer than 120 characters

Tanzhiheng26 commented on src/main/java/chatbot/Alfred.java:
```diff
@@ -0,0 +1,315 @@
+                    } else if (input.startsWith("delete")) {
+                        alfred.deleteList(input.substring(6).trim());
+                    }
+                    else {
```
else should be on the previous line?

Tanzhiheng26 commented on src/main/java/chatbot/Alfred.java:
```diff
@@ -0,0 +1,315 @@
+                    }
+                    input = input.substring(5).trim();
+                    ToDo todo = new ToDo(input);
+                    //addData(todo);
```
Should this be removed?

Tanzhiheng26 commented on src/main/java/chatbot/Alfred.java:
```diff
@@ -0,0 +1,315 @@
+        } catch (FileNotFoundException e) {
+            this.echo("Sorry Master Bruce. I cannot find the data. I will create a new file for you.");
+            // create a new file in the same location
+            file = new java.io.File("data/alfred.txt");
```
I think this creates a new File object but it does not actually create the alfred.txt file? The mkdir() and createNewFile() method in the File class might be what you are looking for

theman-oj10 replied:
Hi when I tested by deleting the alfred.txt and then running my code it works as expected.

Tanzhiheng26 commented on src/main/java/chatbot/Alfred.java:
```diff
@@ -0,0 +1,315 @@
+            file = new java.io.File("data/alfred.txt");
+        }
+    }
+    public void processData(File file) throws FileNotFoundException {
```
Very long method, maybe you can abstract out some parts into functions or add more comments?

theman-oj10 committed:
Alfred Class: Implement Proper Formatting of Dates and Times

Alfred Class stores dates and times as Strings

Storing date and time as Date Time objects will allow for easier parsing, easy comparison and improves code readability.

Let's
* Change all String representation of Dates and Time to LocalDateTime Objects
* Change the current Date and Time parsing to accommodate the LocalDateTime Objects.
* Add additional abstractions to improve code readability.

theman-oj10 committed:
Implement Ui, Storage, Parser and TaskList Classes

Implemented more classes to abstract out all the functionalities further.

Abstracting into these classes will improve code readability, ease of debugging and reduce code complexity.

Let's
* Implement Ui, Storage, Parser, TaskList Classes
* Change Alfred class to now use these subclasses

theman-oj10 committed:
Merge branch 'add-gradle-support'

theman-oj10 committed:
Ui and TaskList Class: Add JUnit Test cases

The package has no unit tests implemented.

Unit testing is required as the code is getting increasingly more complicated.

Let's:
* Implement TaskListTest class with comprehensive test cases
* Implement UiTest class with comprehensive test cases

theman-oj10 committed:
Create v0.1 Release of Alfred

First release in jar format for Alfred.

Initial release will allow for users to run the code and test it, to allow for improvements in next iterations.

Let's
* Create a jar file and publish as release

theman-oj10 committed:
Implement Coding Standard

Fixed certain coding standard violations

Let's:
* Modify IDE settings to enforce coding standards
* Ensure switch statements indentation is correct

theman-oj10 committed:
TaskList Class: Implement Find functionality

Implement ability to find by a certain keyword.

Allows for user to find similar things they have to do and allows for easier categorisation of tasks.

Let's:
* Implement findByKeyword method
* Implement the command 'find' in Alfred Class

theman-oj10 committed:
TaskList Class: Implement Find functionality

Implement ability to find by a certain keyword.

Allows for user to find similar things they have to do and allows for easier categorisation of tasks.

Let's:
* Implement findByKeyword method
* Implement the command 'find' in Alfred Class

theman-oj10 committed:
Implement JavaDocs

No Documentation on the public non-trivial methods.

Lack of documentation affects code readability and code quality.

Let's:
* Add JavaDocs for all Public non-trivial methods

theman-oj10 committed:
Implement a GUI for Alfred

Alfred was using a CLI interface that was not very intuitive to users.

Using a GUI improves user Experience.

Let's
* Implement Alfred GUI, Main Window, DialogBox and other relevant classes for JavaFx
* Fix existing code to work with new implementation.

theman-oj10 committed:
Implement Checkstyle

The coding standard was manually maintained.

Implementing checkstyle configurations will allow for automated detections which will improve developer experience and overall code quality.

Let's:
* Implement checkstyle configurations
* Improve code quality based on suggestions

theman-oj10 committed:
Implement Assertions

The code has no checks to verify assumptions about Internal invariants and relies solely on exception handling.

Implementing assertions can help indicate areas where programmer could have potentially made mistakes.

Let's:
* Implement assertions wherever suitable

theman-oj10 committed:
Improve Code Quality

Code Quality leads to improved code Readability.

Let's:
* Remove arrow-head code
* Implement SLAP

theman-oj10 committed:
Merge pull request #2 from theman-oj10/branch-A-Assertions

Implement Assertions

theman-oj10 committed:
Resolving merge conflicts in branch-A-CodeQuality

theman-oj10 committed:
Merge pull request #3 from theman-oj10/branch-A-CodeQuality

Merging Code Quality Branch to master

theman-oj10 committed:
Implement get urgent tasks feature

The new method returns the most urgent tasks that are not done and sorts deadlines before events.

theman-oj10 committed:
Improve GUI

GUI was improved to add background image, and reformatted text.

theman-oj10 committed:
Add More Testing
Added more automated testing.

theman-oj10 committed:
Add User Guide
Documentation was updated to include updated readme file

theman-oj10 committed:
Fix bugs for release
Minor bugs were fixed and test cases were added.
